src/connectors/snowflake/trulens/connectors/snowflake/utils/server_side_evaluation_artifacts.py

```diff
@@ -0,0 +1,221 @@
+        user: str,
+        database_name: str,
+        schema_name: str,
+        warehouse: str,
+        role: str,
+        database_url: str,
+    ) -> None:
+        self._session = session
+        self._account = account
+        self._user = user
+        self._database_name = database_name
+        self._schema_name = schema_name
+        self._warehouse = warehouse
+        self._role = role
+        self._database_url = database_url
+        self._validate_name(database_name, "database_name")
+        self._validate_name(schema_name, "schema_name")
+        self._validate_name(warehouse, "warehouse")
+
+    @staticmethod
+    def _validate_name(name: str, error_message_variable_name: str) -> None:
+        if not re.match(r"^[A-Za-z0-9_]+$", name):
+            raise ValueError(
+                f"`{error_message_variable_name}` must contain only alphanumeric and underscore characters!"
+            )
+
+    def set_up_all(self) -> None:
+        self._set_up_stage()
+        self._set_up_stream()
+        self._set_up_stored_procedure()
+        self._set_up_wrapper_stored_procedure()
+        self._set_up_task()
+
+    @staticmethod
+    def _remove_leading_spaces(q: str) -> str:
+        # Get leading whitespace in first non-empty line.
+        length_of_leading_whitespace = 0
+        leading_whitespace_in_first_non_empty_line = ""
+        for line in q.split("\n"):
+            if line.strip():
+                length_of_leading_whitespace = len(line) - len(line.lstrip())
+                leading_whitespace_in_first_non_empty_line = line[
+                    :length_of_leading_whitespace
+                ]
+                break
+        # Remove the leading whitespace in the first non-empty line from all non-empty lines.
+        ret = []
+        for line in q.split("\n"):
+            if line.startswith(leading_whitespace_in_first_non_empty_line):
+                ret.append(line[length_of_leading_whitespace:])
+            elif not line.strip():
+                ret.append("")
+            else:
+                raise ValueError()
+        return "\n".join(ret).strip()
+
+    def _run_query(self, q: str) -> None:
+        # Before running the query we first remove leading spaces for two reasons:
+        # 1. Python code in a stored procedure needs to have correct indenting.
+        # 2. To format the query for easier readability in debugging scenarios.
+        self._session.sql(self._remove_leading_spaces(q)).collect()
+
+    def _set_up_stage(self) -> None:
+        self._run_query(f"CREATE STAGE IF NOT EXISTS {_STAGE_NAME}")
+        data_directory = os.path.join(
+            os.path.dirname(__file__), "../../../data/snowflake_stage_zips"
+        )
+        for zip_to_upload in _ZIPS_TO_UPLOAD:
+            self._session.file.put(
+                os.path.join(data_directory, zip_to_upload),
+                f"@{_STAGE_NAME}",
+            )
+
+    def _set_up_stream(self) -> None:
+        self._run_query(
```

Comment:
out of curiosity, why don't we use snowpark for setting up the stream/stage/proc?

Reply:
mostly because I first tested things out via sql and then turned it into python. If you think it's much nicer via snowpark though I can look into it.